File: C++/minimum-operations-to-make-a-uni-value-grid.cpp
// Time:  O(m * n) on average
// Space: O(m * n)

class Solution {
public:
    int minOperations(vector<vector<int>>& grid, int x) {
        vector<int> nums;
        for (const auto& row : grid) {
            nums.insert(end(nums), cbegin(row), cend(row));
        }
        int mod_group = -1;
        for (const auto& num : nums) {
            if (mod_group == -1) {
                mod_group = num % x;
            } else if (mod_group != num % x) {
                return -1;
            }
        }
        nth_element(begin(nums), begin(nums) + size(nums) / 2, end(nums));
        const auto& median = nums[size(nums) / 2];
        return accumulate(cbegin(nums), cend(nums), 0,
                          [&median, &x](const auto& total, const auto& a) {
                              return total + abs(a - median) / x;
                          });
    }
};

File: Python/minimum-operations-to-make-a-uni-value-grid.py
# Time:  O(m * n) on average
# Space: O(m * n)

import random

class Solution(object):
    def minOperations(self, grid, x):
        """
        :type grid: List[List[int]]
        :type x: int
        :rtype: int
        """
        def nth_element(nums, n, compare=lambda a, b: a < b):
            def tri_partition(nums, left, right, target, compare):
                mid = left
                while mid <= right:
                    if nums[mid] == target:
                        mid += 1
                    elif compare(nums[mid], target):
                        nums[left], nums[mid] = nums[mid], nums[left]
                        left += 1
                        mid += 1
                    else:
                        nums[mid], nums[right] = nums[right], nums[mid]
                        right -= 1
                return left, right

            left, right = 0, len(nums)-1
            while left <= right:
                pivot_idx = random.randint(left, right)
                pivot_left, pivot_right = tri_partition(nums, left, right, nums[pivot_idx], compare)
                if pivot_left <= n <= pivot_right:
                    return
                elif pivot_left > n:
                    right = pivot_left-1
                else:  # pivot_right < n.
                    left = pivot_right+1

        nums = [v for row in grid for v in row]
        if len(set(v%x for v in nums)) > 1:
            return -1
        nth_element(nums, len(nums)//2)
        median = nums[len(nums)//2]
        return sum(abs(v-median)//x for v in nums)
